### PR TITLE
Notebook Instance lifecycle configuration (with samples)

### PR DESCRIPTION
**Pull request by ananth102**
This is a pull request for the Notebook Instance Lifecycle configuration and its samples. It contains all crud operations.

Files added:

custom_delta.go: This file contains two functions that return an updated delta if the onCreate/onStart values have changed.
Both delta functions do a nil check, length check and finally check if the contents of onStart/onCreate have changed. The code generator does not support this natively so it has to be written by hand for now.

Note: onCreate and onStart have identical structures thus both delta functions are the same except for different variable names.

delta_post_compare.go.tpl: hook that goes in delta.go that calls the functions in custom_delta.go

my-notebookinstance_lifecycle_config.yaml - Sample Notebook Instance Lifecycle configuration

Changed files:

generator.yaml: Added the Notebook Lifecycle configuration and a hook for delta. Also added a field called LastModifiedTime to inform the user about the last time the resource was modified.

### CODE REVIEW

File: generator.yaml [hunk 1]
```diff
@@ -22,6 +22,14 @@ operations:
   StopHyperParameterTuningJob:
     operation_type: Delete
     resource_name: HyperParameterTuningJob
+  DescribeModelQualityJobDefinition:
```
Comment (ryansteakley) [inline generator.yaml:25]: Believe this was already added by previous pr's ?

File: pkg/resource/notebook_instance_lifecycle_config/manager_factory.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	rm, err := newResourceManager(cfg, log, metrics, rr, sess, id, region)
+	if err != nil {
+		return nil, err
+	}
+	f.rmCache[rmId] = rm
+	return rm, nil
+}
+
+// IsAdoptable returns true if the resource is able to be adopted
+func (f *resourceManagerFactory) IsAdoptable() bool {
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance_lifecycle_config/manager_factory.go:78]: Have you tested/verified that it works properly.

File: templates/notebook_instance_lifecycle_config/delta_post_compare.go.tpl [hunk 3]
```diff
@@ -0,0 +1,2 @@
+delta = modifyDeltaCreate(delta,a,b)
```
Comment (ryansteakley) [inline templates/notebook_instance_lifecycle_config/delta_post_compare.go.tpl:1]: nit: Can you possibly format this file ex: delta = modifyDeltaCreate(delta, a, b)

File: samples/notebookinstance_lifecycle_config/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+This command lists Notebook Instance Lifecycle Configurations and their last modified times.
+```
+$ kubectl get NotebookInstanceLifecycleConfig 
+```
+### Delete a Notebook Instance Lifecycle Configuration
+This command deletes a Notebook Instance Lifecycle Configuration.
+```
+$ kubectl delete NotebookInstanceLifecycleConfig my-notebook_lifecycle_config
+```
+
```
Comment (ryansteakley) [inline samples/notebookinstance_lifecycle_config/README.md:34]: nit: Multiple extra new lines

**Review by ryansteakley — COMMENTED** [review]
Will there be tests coming to this pr?

File: pkg/resource/notebook_instance_lifecycle_config/manager_factory.go (hunk 2 above)
Comment (ananth102) [reply pkg/resource/notebook_instance_lifecycle_config/manager_factory.go:78]: Yea adopted resources seem to work just fine.

File: pkg/resource/notebook_instance_lifecycle_config/custom_delta.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		if len(a.ko.Spec.OnCreate) != len(b.ko.Spec.OnCreate) {
+			delta.Add("Spec.OnCreate", a.ko.Spec.OnCreate, b.ko.Spec.OnCreate)
+			return delta
+		}
+	} else {
+		return delta
+	}
+	//Check variables, a and b have to be of equal length
+	onCreateLen := len(a.ko.Spec.OnCreate)
+	if a.ko.Spec.OnCreate != nil && b.ko.Spec.OnCreate != nil {
```
Comment (mbaijal) [inline pkg/resource/notebook_instance_lifecycle_config/custom_delta.go:27]: is it possible to test this using the update test ? Else we should make sure to add it to unit tests.

**Review by mbaijal — APPROVED** [review]

**Pull request closed without merging**